lib/Frontend/CompilerInvocation.cpp

```diff
@@ -247,637 +246,758 @@ class ArgsToFrontendInputsConverter {
+}
+
+bool FrontendArgsToOptionsConverter::setupForSILOrLLVM() {
+  using namespace options;
+  bool TreatAsSIL =
+  Args.hasArg(OPT_parse_sil) || Opts.Inputs.shouldTreatAsSIL();
+  bool TreatAsLLVM = Opts.Inputs.shouldTreatAsLLVM();
+  
+  if (Opts.Inputs.verifyInputs(
+                               Diags, TreatAsSIL, Opts.RequestedAction == FrontendOptions::REPL,
+                               Opts.RequestedAction == FrontendOptions::NoneAction)) {
+    return true;
+  }
+  if (Opts.RequestedAction == FrontendOptions::Immediate) {
+    Opts.ImmediateArgv.push_back(
+                                 Opts.Inputs.getFilenameOfFirstInput()); // argv[0]
+    if (const Arg *A = Args.getLastArg(OPT__DASH_DASH)) {
+      for (unsigned i = 0, e = A->getNumValues(); i != e; ++i) {
+        Opts.ImmediateArgv.push_back(A->getValue(i));
+      }
+    }
+  }
+  
+  if (TreatAsSIL)
+    Opts.InputKind = InputFileKind::IFK_SIL;
+  else if (TreatAsLLVM)
+    Opts.InputKind = InputFileKind::IFK_LLVM_IR;
+  else if (Args.hasArg(OPT_parse_as_library))
+    Opts.InputKind = InputFileKind::IFK_Swift_Library;
+  else if (Opts.RequestedAction == FrontendOptions::ActionType::REPL)
+    Opts.InputKind = InputFileKind::IFK_Swift_REPL;
+  else
+    Opts.InputKind = InputFileKind::IFK_Swift;
+  
+  return false;
+}
+
+void FrontendArgsToOptionsConverter::setModuleName() {
+  const Arg *A = Args.getLastArg(options::OPT_module_name);
+  if (A) {
+    Opts.ModuleName = A->getValue();
+  } else if (Opts.ModuleName.empty()) {
+    // The user did not specify a module name, so determine a default fallback
+    // based on other options.
+    
+    // Note: this code path will only be taken when running the frontend
+    // directly; the driver should always pass -module-name when invoking the
+    // frontend.
+    Opts.ModuleName = determineFallbackModuleName();
+  }
+  
+  if (Lexer::isIdentifier(Opts.ModuleName) &&
+      (Opts.ModuleName != STDLIB_NAME || Opts.ParseStdlib)) {
+    return;
+  }
+  if (!Opts.actionHasOutput() || Opts.isCompilingExactlyOneSwiftFile()) {
+    Opts.ModuleName = "main";
+    return;
+  }
+  auto DID = (Opts.ModuleName == STDLIB_NAME) ? diag::error_stdlib_module_name
+  : diag::error_bad_module_name;
+  Diags.diagnose(SourceLoc(), DID, Opts.ModuleName, A == nullptr);
+  Opts.ModuleName = "__bad__";
+}
+
+StringRef FrontendArgsToOptionsConverter::determineFallbackModuleName() const {
+  // Note: this code path will only be taken when running the frontend
+  // directly; the driver should always pass -module-name when invoking the
+  // frontend.
+  if (Opts.RequestedAction == FrontendOptions::REPL) {
+    // Default to a module named "REPL" if we're in REPL mode.
+    return "REPL";
+  }
+  // In order to pass Driver/options.swift test must leave ModuleName empty
+  if (!Opts.Inputs.hasInputFilenames()) {
+    return StringRef();
+  }
+  std::vector<std::string> outputFilenames = getUnprocessedOutputFilenames();
```

Comment:
I'm not happy with this (even though it'll only come up in tests) because it means more file I/O and more in-memory copies. Haven't we loaded the output filenames by this point? Can't we just grab them from wherever they usually live?